// @vkontakte/vkui/dist/components/SelectMimicry/SelectMimicry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classNames2 = _interopRequireDefault(require("../../lib/classNames"));

var _dropdown = _interopRequireDefault(require("@vkontakte/icons/dist/24/dropdown"));

var _FormField = _interopRequireDefault(require("../FormField/FormField"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var SelectMimicry = function SelectMimicry(_ref) {
  var className = _ref.className,
      tabIndex = _ref.tabIndex,
      placeholder = _ref.placeholder,
      children = _ref.children,
      alignment = _ref.alignment,
      status = _ref.status,
      getRootRef = _ref.getRootRef,
      multiline = _ref.multiline,
      disabled = _ref.disabled,
      restProps = _objectWithoutProperties(_ref, ["className", "tabIndex", "placeholder", "children", "alignment", "status", "getRootRef", "multiline", "disabled"]);

  return _react.default.createElement(_FormField.default, _extends({}, restProps, {
    tabIndex: disabled ? null : tabIndex,
    className: (0, _classNames2.default)('Select', 'Select--mimicry', _defineProperty({
      'Select--not-selected': !children,
      'Select--multiline': multiline,
      'Select--disabled': disabled
    }, "Select--align-".concat(alignment), alignment), className),
    getRootRef: getRootRef,
    status: status
  }), _react.default.createElement("div", {
    className: "Select__container"
  }, _react.default.createElement("div", {
    className: "Select__title"
  }, children || placeholder), _react.default.createElement(_dropdown.default, null)));
};

SelectMimicry.propTypes = {
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  children: _propTypes.default.node,
  tabIndex: _propTypes.default.number,
  placeholder: _propTypes.default.string,
  alignment: _propTypes.default.oneOf(['left', 'center', 'top']),
  getRootRef: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.shape({
    current: _propTypes.default.instanceOf(Element)
  })]),
  multiline: _propTypes.default.bool,
  disabled: _propTypes.default.bool,
  status: _propTypes.default.oneOf(['default', 'error', 'valid'])
};
SelectMimicry.defaultProps = {
  tabIndex: 0
};
var _default = SelectMimicry;
exports.default = _default;
