// @vkontakte/vkui/dist/components/Header/Header.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _getClassName = _interopRequireDefault(require("../../helpers/getClassName"));

var _classNames2 = _interopRequireDefault(require("../../lib/classNames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var baseClassNames = (0, _getClassName.default)('Header');

var Header = function Header(_ref) {
  var className = _ref.className,
      level = _ref.level,
      children = _ref.children,
      indicator = _ref.indicator,
      aside = _ref.aside,
      getRootRef = _ref.getRootRef,
      restProps = _objectWithoutProperties(_ref, ["className", "level", "children", "indicator", "aside", "getRootRef"]);

  return _react.default.createElement("div", _extends({}, restProps, {
    ref: getRootRef,
    className: (0, _classNames2.default)(baseClassNames, className, _defineProperty({}, "Header--level-".concat(level), true))
  }), _react.default.createElement("div", {
    className: "Header__in"
  }, _react.default.createElement("div", {
    className: "Header__content"
  }, children), indicator && _react.default.createElement("div", {
    className: "Header__indicator"
  }, indicator), aside && _react.default.createElement("div", {
    className: "Header__aside"
  }, aside)));
};

Header.propTypes = {
  className: _propTypes.default.string,
  level: _propTypes.default.oneOf(['1', '2']),
  indicator: _propTypes.default.node,
  aside: _propTypes.default.node,
  children: _propTypes.default.node,
  style: _propTypes.default.object,
  getRootRef: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.shape({
    current: _propTypes.default.instanceOf(Element)
  })])
};
Header.defaultProps = {
  level: '1'
};
var _default = Header;
exports.default = _default;
